File: contents/verlet_integration/code/python/verlet.py
def verlet(pos, acc, dt):
    prev_pos = pos
    time = 0

    while pos > 0:
        time += dt
        next_pos = pos * 2 - prev_pos + acc * dt * dt
        prev_pos, pos = pos, next_pos

    return time

def stormer_verlet(pos, acc, dt):
    prev_pos = pos
    time = 0
    vel = 0

    while pos > 0:
        time += dt
        next_pos = pos * 2 - prev_pos + acc * dt * dt
        prev_pos, pos = pos, next_pos
        vel += acc * dt

    return time, vel

def velocity_verlet(pos, acc, dt):
    time = 0
    vel = 0

    while pos > 0:
        time += dt
        pos += vel * dt + 0.5 * acc * dt * dt
        vel += acc * dt

    return time, vel

def main():
    time = verlet(5, -10, 0.01)
    print("[#]\nTime for Verlet integration is:")
    print("{:.10f}".format(time))

    time, vel = stormer_verlet(5, -10, 0.01)
    print("[#]\nTime for Stormer Verlet integration is:")
    print("{:.10f}".format(time))
    print("[#]\nVelocity for Stormer Verlet integration is:")
    print("{:.10f}".format(vel))

    time, vel = velocity_verlet(5, -10, 0.01)
    print("[#]\nTime for velocity Verlet integration is:")
    print("{:.10f}".format(time))
    print("[#]\nVelocity for velocity Verlet integration is:")
    print("{:.10f}".format(vel))


if __name__ == '__main__':
    main()

File: contents/verlet_integration/code/nim/verlet.nim
func verlet(pos_in, acc, dt: float): float =
  var
    pos: float = pos_in
    prevPos: float = pos
    time: float = 0.0
    tempPos: float

  while pos > 0.0:
    time += dt
    tempPos = pos
    pos = pos * 2 - prevPos + acc * dt * dt
    prevPos = tempPos

  time

func stormerVerlet(pos_in, acc, dt: float): (float, float) =
  var
    pos: float = pos_in
    prevPos: float = pos
    time: float = 0.0
    vel: float = 0.0
    tempPos: float

  while pos > 0.0:
    time += dt
    tempPos = pos
    pos = pos * 2 - prevPos + acc * dt * dt
    prevPos = tempPos

    vel += acc * dt

  (time, vel)

func velocityVerlet(pos_in, acc, dt: float): (float, float) =
  var
    pos: float = pos_in
    time: float = 0.0
    vel: float = 0.0

  while pos > 0.0:
    time += dt
    pos += vel * dt + 0.5 * acc * dt * dt
    vel += acc * dt

  (time, vel)

when isMainModule:
  let timeV = verlet(5.0, -10.0, 0.01)
  echo "[#]\nTime for Verlet integration is:"
  echo timeV

  let (timeSV, velSV) = stormerVerlet(5.0, -10.0, 0.01)
  echo "[#]\nTime for Stormer Verlet integration is:"
  echo timeSV
  echo "[#]\nVelocity for Stormer Verlet integration is:"
  echo velSV

  let (timeVV, velVV) = velocityVerlet(5.0, -10.0, 0.01)
  echo "[#]\nTime for velocity Verlet integration is:"
  echo timeVV
  echo "[#]\nVelocity for velocity Verlet integration is:"
  echo velVV
